tests/core/training/story_reader/test_common_story_reader.py

```diff
@@ -0,0 +1,258 @@
+    assert tracker.events[3] == ActionExecuted("utter_greet")
+    assert tracker.events[4] == ActionExecuted("action_listen")
+
+
+@pytest.mark.parametrize(
+    "stories_file",
+    [
+        "data/test_stories/stories_checkpoint_after_or.md",
+        "data/test_yaml_stories/stories_checkpoint_after_or.yml",
+    ],
+)
+async def test_can_read_test_story_with_checkpoint_after_or(
+    stories_file: Text, default_domain
+):
+    trackers = await training.load_data(
+        stories_file,
+        default_domain,
+        use_story_concatenation=False,
+        tracker_limit=1000,
+        remove_duplicates=False,
+    )
+    # there should be only 2 trackers
+    assert len(trackers) == 2
+
+
+@pytest.mark.parametrize(
+    "stories_file",
+    [
+        "data/test_stories/stories_with_cycle.md",
+        "data/test_yaml_stories/stories_with_cycle.yml",
+    ],
+)
+async def test_read_story_file_with_cycles(stories_file: Text, tmpdir, default_domain):
+    graph = await training.extract_story_graph(stories_file, default_domain)
+
+    assert len(graph.story_steps) == 5
+
+    graph_without_cycles = graph.with_cycles_removed()
+
+    assert graph.cyclic_edge_ids != set()
+    # sorting removed_edges converting set converting it to list
+    assert graph_without_cycles.cyclic_edge_ids == list()
+
+    assert len(graph.story_steps) == len(graph_without_cycles.story_steps) == 5
+
+    assert len(graph_without_cycles.story_end_checkpoints) == 2
+
+
+@pytest.mark.parametrize(
+    "stories_file",
+    [
+        "data/test_stories/stories_with_cycle.md",
+        "data/test_yaml_stories/stories_with_cycle.yml",
+    ],
+)
+async def test_generate_training_data_with_cycles(stories_file: Text, default_domain):
+    featurizer = MaxHistoryTrackerFeaturizer(
+        BinarySingleStateFeaturizer(), max_history=4
+    )
+    training_trackers = await training.load_data(
+        stories_file, default_domain, augmentation_factor=0
+    )
+    training_data = featurizer.featurize_trackers(training_trackers, default_domain)
+    y = training_data.y.argmax(axis=-1)
+
+    # how many there are depends on the graph which is not created in a
+    # deterministic way but should always be 3 or 4
+    assert len(training_trackers) == 3 or len(training_trackers) == 4
+
+    # if we have 4 trackers, there is going to be one example more for label 10
+    num_tens = len(training_trackers) - 1
+    # if new default actions are added the keys of the actions will be changed
+
+    assert Counter(y) == {0: 6, 10: num_tens, 12: 1, 1: 2, 11: 3}
+
+
+@pytest.mark.parametrize(
+    "stories_file",
+    [
+        "data/test_stories/stories_unused_checkpoints.md",
+        "data/test_yaml_stories/stories_unused_checkpoints.yml",
+    ],
+)
+async def test_generate_training_data_with_unused_checkpoints(
+    stories_file: Text, tmpdir, default_domain
+):
+    training_trackers = await training.load_data(stories_file, default_domain)
```

Comment:
should we test for a warning then?

Reply:
I'm not sure, that's not incorrect isn't it?